[symm_mem] Move all symm mem code into a dedicated folder

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #155573

We arrive at a point when so many files are related to symmetric memory and files are scattered around in the cpp side. Let's first put all related code (symmetric memory related) into a separate folder. We can do further refactoring later if needed.

cc @H-Huang @awgu @wanchaol @fegin @wz337 @wconstab @d4l3k

fduwjj committed:
[symm_mem] Move all symm mem code into a dedicated folder

[ghstack-poisoned]

Skylion007 commented on torch/csrc/distributed/c10d/symm_mem/intra_node_comm.hpp:
```diff
@@ -3,7 +3,7 @@
 #include <c10/cuda/CUDAStream.h>
 #include <torch/csrc/distributed/c10d/Store.hpp>
-#include <torch/csrc/distributed/c10d/SymmetricMemory.hpp>
+#include <torch/csrc/distributed/c10d/symm_mem/SymmetricMemory.hpp>
```
Can we apply clang-tidy to this entire subfolder in .lintrunner? <3 (optin clang-tidy / clang-format if the relevant headers pass)

fduwjj replied:
sure, it is still draft. I am update the PR right now.

fduwjj committed:
Update on "[symm_mem] Move all symm mem code into a dedicated folder"




cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]

fduwjj committed:
Update on "[symm_mem] Move all symm mem code into a dedicated folder"


We arrive at a point when so many files are related to symmetric memory and files are scattered around in the cpp side. Let's first put all related code (symmetric memory related) into a separate folder. We can do further refactoring later if needed.

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]

fegin reviewed (approved):
stamp to unblock

d4l3k reviewed (approved):
LGTM
